### PR TITLE
[#5382] Ensure the underscore locale is included in i18n fallbacks

### PR DESCRIPTION
**Pull request by garethrees**
## Relevant issue(s)

Fixes https://github.com/mysociety/alaveteli/issues/5382

## What does this do?

* Significant cleanup of `AlaveteliLocalization`
* Ensure the underscore locale is included in i18n fallbacks

## Why was this needed?

Lack of correct fallback results in calling methods on `nil` where we're expecting a translated attribute.

## Notes to reviewer

There are a few public methods where we're not yet parsing the values into `AlaveteliLocalization::Locale` classes. I think it would be nice to do this, but there's probably wider-reaching impact, so perhaps something for another day.

### CODE REVIEW

File: spec/lib/alaveteli_localization_spec.rb [hunk 1]
```diff
@@ -60,6 +60,19 @@
         expect(I18n.available_locales).to eq([:"en-GB", :en, :es])
       end
 
+      it 'sets correct fallbacks when using an underscore default locale' do
+        AlaveteliLocalization.set_locales('en_RW rw', 'en_RW')
+        expect(I18n.fallbacks[:en_RW]).to eq([:en_RW, :en])
+        expect(I18n.fallbacks[:rw]).to eq([:rw, :en, 'en_RW'])
```
Comment (garethrees) [inline spec/lib/alaveteli_localization_spec.rb:66]: `[:rw, 'en_RW', en]` would be preferable.

**Review by gbp — CHANGES_REQUESTED** [review]
This errors for me, not sure why it's working on CI.

`I18n.fallbacks` isn't avalible until: https://github.com/mysociety/alaveteli/blob/35219f1765dd2c9ae47d19aaab35288de255f9f3/config/initializers/fast_gettext.rb#L15 but before this initializer in run `AlaveteliLocalization` is loaded.

We need something like:

```diff
diff --git a/config/initializers/fast_gettext.rb b/config/initializers/fast_gettext.rb
index af5f9c8d9b..4709984913 100644
--- a/config/initializers/fast_gettext.rb
+++ b/config/initializers/fast_gettext.rb
@@ -11,7 +11,4 @@ if AlaveteliConfiguration::enable_alaveteli_pro
   repos << pro_repo
 end
 AlaveteliLocalization.set_default_text_domain('app', repos)
-
-I18n::Backend::Simple.send(:include, I18n::Backend::Fallbacks)
-
 AlaveteliLocalization.set_default_locale_urls(AlaveteliConfiguration::include_default_locale_in_urls)
diff --git a/lib/alaveteli_localization.rb b/lib/alaveteli_localization.rb
index 2dea217c71..e8d6f00714 100644
--- a/lib/alaveteli_localization.rb
+++ b/lib/alaveteli_localization.rb
@@ -1,4 +1,7 @@
 # -*- encoding : utf-8 -*-
+
+I18n::Backend::Simple.send(:include, I18n::Backend::Fallbacks)
+
 class AlaveteliLocalization
   class << self
     def set_locales(available_locales, default_locale)
```

**Comment by garethrees** [comment]
> This errors for me

I don't get an error – could you paste stack trace and your `DEFAULT_LOCALE` / `AVAILABLE_LOCALE` values?

**Comment by garethrees** [comment]
We've just discovered that even though the specs pass, the `Globalize.fallbacks` do not persist when booting in the app console with an underscore default locale.

The following patch makes these fallbacks persist, but we're pretty sure that Globalize should be [inheriting fallbacks](https://github.com/globalize/globalize/blob/v5.2.0/lib/globalize.rb#L53) from I18n, so maybe there's a load order problem going on here, where I18n and/or Globalize is being initialized before AlaveteliLocalization.

```diff
diff --git a/lib/alaveteli_localization.rb b/lib/alaveteli_localization.rb
index f3a6199d0..d82ee10b3 100644
--- a/lib/alaveteli_localization.rb
+++ b/lib/alaveteli_localization.rb
@@ -10,6 +10,7 @@ class AlaveteliLocalization
 
       set_fast_gettext_locales(available, default)
       set_i18n_locales(available, default)
+      set_globalize_locales(available, default)
       set_conditionally_prepend_locale_locales(available, default)
     end
 
@@ -75,6 +76,15 @@ class AlaveteliLocalization
       FastGettext.default_locale = default.canonicalize.to_s
     end
 
+    def set_globalize_locales(available, default)
+       fallbacks = {
+        fr_BE: [:fr_BE, :"fr-BE", :fr],
+        nl_BE: [:nl_BE, :fr_BE, :"fr-BE", :fr]
+      }
+
+      Globalize.fallbacks = fallbacks
+    end
+
     def set_i18n_locales(available, default)
       i18n_locales = available.each_with_object([]) do |locale, memo|
         memo.concat(locale.self_and_parents)
@@ -86,6 +96,7 @@ class AlaveteliLocalization
     end
 
     def set_i18n_fallbacks(available, default)
+      I18n::Backend::Simple.send(:include, I18n::Backend::Fallbacks)
       return unless default.is_a?(UnderscorredLocale)
       available.each { |locale| add_i18n_fallback(default, locale) }
     end
```

**Comment by garethrees** [comment]
Strange. When hard-coding `Globalize.fallbacks` as above, the app boots with the correct fallbacks. Here's some debugging info to show the fallbacks at each translation-related initializer.

```diff
diff --git a/config/initializers/alaveteli.rb b/config/initializers/alaveteli.rb
index 09cd7085b..595413526 100644
--- a/config/initializers/alaveteli.rb
+++ b/config/initializers/alaveteli.rb
@@ -70,6 +70,9 @@ require 'alaveteli_pro/webhook_endpoints'
 AlaveteliLocalization.set_locales(AlaveteliConfiguration::available_locales,
                                   AlaveteliConfiguration::default_locale)
 
+
+puts "==== init/alaveteli \n\n#{Globalize.fallbacks('nl_BE')}\n\n"
+
 # Allow tests to be run under a non-superuser database account if required
 if Rails.env == 'test' and ActiveRecord::Base.configurations['test']['constraint_disabling'] == false
   require 'no_constraint_disabling'
diff --git a/config/initializers/fast_gettext.rb b/config/initializers/fast_gettext.rb
index af5f9c8d9..d3f6f699a 100644
--- a/config/initializers/fast_gettext.rb
+++ b/config/initializers/fast_gettext.rb
@@ -15,3 +15,5 @@ AlaveteliLocalization.set_default_text_domain('app', repos)
 I18n::Backend::Simple.send(:include, I18n::Backend::Fallbacks)
 
 AlaveteliLocalization.set_default_locale_urls(AlaveteliConfiguration::include_default_locale_in_urls)
+
+puts "==== init/fast_gettext \n\n#{Globalize.fallbacks('nl_BE')}\n\n"
diff --git a/config/initializers/gettext_i18n_rails.rb b/config/initializers/gettext_i18n_rails.rb
index 3beb2e58b..923fbc4e7 100644
--- a/config/initializers/gettext_i18n_rails.rb
+++ b/config/initializers/gettext_i18n_rails.rb
@@ -2,3 +2,5 @@
 # FIXME: Audit the translations for XSS opportunities. Ultimately it would be
 # good to get rid of this and explicitly mark strings as html_safe
 GettextI18nRails.translations_are_html_safe = true
+
+puts "==== init/gettext_i18n_rails \n\n#{Globalize.fallbacks('nl_BE')}\n\n"
diff --git a/lib/alaveteli_localization.rb b/lib/alaveteli_localization.rb
index f3a6199d0..d82ee10b3 100644
--- a/lib/alaveteli_localization.rb
+++ b/lib/alaveteli_localization.rb
@@ -10,6 +10,7 @@ class AlaveteliLocalization
 
       set_fast_gettext_locales(available, default)
       set_i18n_locales(available, default)
+      set_globalize_locales(available, default)
       set_conditionally_prepend_locale_locales(available, default)
     end
 
@@ -75,6 +76,15 @@ class AlaveteliLocalization
       FastGettext.default_locale = default.canonicalize.to_s
     end
 
+    def set_globalize_locales(available, default)
+       fallbacks = {
+        fr_BE: [:fr_BE, :"fr-BE", :fr],
+        nl_BE: [:nl_BE, :fr_BE, :"fr-BE", :fr]
+      }
+
+      Globalize.fallbacks = fallbacks
+    end
+
     def set_i18n_locales(available, default)
       i18n_locales = available.each_with_object([]) do |locale, memo|
         memo.concat(locale.self_and_parents)
@@ -86,6 +96,7 @@ class AlaveteliLocalization
     end
 
     def set_i18n_fallbacks(available, default)
+      I18n::Backend::Simple.send(:include, I18n::Backend::Fallbacks)
       return unless default.is_a?(UnderscorredLocale)
       available.each { |locale| add_i18n_fallback(default, locale) }
     end
```

Console session showing fallbacks persisted on boot:

```
$ bin/rails c
==== init/alaveteli

[:nl_BE, :fr_BE, :"fr-BE", :fr]

==== init/fast_gettext

[:nl_BE, :fr_BE, :"fr-BE", :fr]

==== init/gettext_i18n_rails

[:nl_BE, :fr_BE, :"fr-BE", :fr]

Loading development environment (Rails 5.2.4.4)
Globalize.fallbacks('nl_BE')
=> [:nl_BE, :fr_BE, :"fr-BE", :fr]
```

However, when booting the rails server and accessing a body with no translations for the given locale the fallbacks are _not_ persisted.

Screenshot 2020-09-24 at 10 17 57

It looks like we're actually losing the fallbacks as we use `AlaveteliLocalization.with_locale`:

```
$ bin/rails c
==== init/alaveteli

[:nl_BE, :fr_BE, :"fr-BE", :fr]

==== init/fast_gettext

[:nl_BE, :fr_BE, :"fr-BE", :fr]

==== init/gettext_i18n_rails

[:nl_BE, :fr_BE, :"fr-BE", :fr]

Loading development environment (Rails 5.2.4.4)

Globalize.fallbacks('nl_BE')
=> [:nl_BE, :fr_BE, :"fr-BE", :fr]

AlaveteliLocalization.with_locale('nl_BE') do
  b = PublicBody.find_by(url_name: 'example_body_1004')
  puts "nl_BE name: #{b.name}"
end

PublicBody Load (2.0ms)  SELECT  DISTINCT "public_bodies"."id", "public_bodies"."version", "public_bodies"."last_edit_editor", "public_bodies"."last_edit_comment", "public_bodies"."created_at", "public_bodies"."updated_at", "public_bodies"."home_page", "public_bodies"."api_key", "public_bodies"."info_requests_count", "public_bodies"."disclosure_log", "public_bodies"."info_requests_successful_count", "public_bodies"."info_requests_not_held_count", "public_bodies"."info_requests_overdue_count", "public_bodies"."info_requests_visible_classified_count", "public_bodies"."info_requests_visible_count" FROM "public_bodies" INNER JOIN "public_body_translations" ON "public_body_translations"."public_body_id" = "public_bodies"."id" WHERE "public_body_translations"."url_name" = $1 AND "public_body_translations"."locale" IN ($2, $3, $4, $5) LIMIT $6  [["url_name", "example_body_1004"], ["locale", "nl_BE"], ["locale", "fr_BE"], ["locale", "fr-BE"], ["locale", "fr"], ["LIMIT", 1]]

NoMethodError: undefined method `name' for nil:NilClass
from (pry):4:in `block in <main>'

```

**Comment by garethrees** [comment]
Looks like the `I18n` fallbacks aren't getting set during the initialization process. They do become available for the console (less the underscore default) but not clear where that happens

```
$ bin/rails c
==== init/alaveteli.rb

GLOBALIZE [:nl_BE, :fr_BE, :"fr-BE", :fr]
I18N [:nl_BE]

==== init/fast_gettext.rb

GLOBALIZE [:nl_BE, :fr_BE, :"fr-BE", :fr]
I18N [:nl_BE]

==== init/gettext_i18n_rails.rb

GLOBALIZE [:nl_BE, :fr_BE, :"fr-BE", :fr]
I18N [:nl_BE]

Loading development environment (Rails 5.2.4.4)
[1](main)> I18n.fallbacks['nl_BE']
=> [:nl_BE, :"fr-BE", :fr]
[2](main)> Globalize.fallbacks('nl_BE')
=> [:nl_BE, :fr_BE, :"fr-BE", :fr]
```

**Comment by garethrees** [comment]
If we call `AlaveteliLocalization.set_locales` at the start of a console session we get the correct behaviour (though the `I18n.fallbacks` array still doesn't include the fallback we want).

```ruby
[1](main)> AlaveteliLocalization.set_locales 'fr_BE nl_BE', 'fr_BE'
=> ["fr_BE", "nl_BE"]

[2](main)> AlaveteliLocalization.with_locale('nl_BE') do
[2](main)*   b = PublicBody.find_by(url_name: 'example_body_1004')
[2](main)*   puts "nl_BE name: #{b.name}"
[2](main)* end

==== AL.with_locale before

GLOBALIZE [:nl_BE, :fr_BE, :"fr-BE", :fr]
I18N [:nl_BE, :"fr-BE", :fr]

PublicBody Load (2.1ms)  SELECT  DISTINCT [snip…]

nl_BE name: Example Public Body 1004

==== AL.with_locale after

GLOBALIZE [:nl_BE, :fr_BE, :"fr-BE", :fr]
I18N [:nl_BE, :"fr-BE", :fr]
```

**Comment by garethrees** [comment]
Adding `config/initializers/Z_test.rb` with the following contents gives us the [behaviour above](https://github.com/mysociety/alaveteli/pull/5893#issuecomment-698263811). Seems like we have a load-order problem.

```ruby
require 'routing_filters'
require 'alaveteli_localization'

AlaveteliLocalization.set_locales(AlaveteliConfiguration::available_locales,
                                  AlaveteliConfiguration::default_locale)
```

**Comment by gbp** [comment]
With 6c9de7e following specs pass

```diff
diff --git a/spec/lib/alaveteli_localization_spec.rb b/spec/lib/alaveteli_localization_spec.rb
index 469437a52c..132f35cbef 100644
--- a/spec/lib/alaveteli_localization_spec.rb
+++ b/spec/lib/alaveteli_localization_spec.rb
@@ -226,4 +226,37 @@ describe AlaveteliLocalization do
   end
+  describe 'Globalize model attributes fallbacks' do
+    before do
+      AlaveteliLocalization.set_locales('en_RW rw', 'en_RW')
+    end
+
+    let(:body_with_default_locale) do
+      AlaveteliLocalization.with_locale('en_RW') do
+        FactoryBot.create(:public_body, url_name: 'default')
+      end
+    end
+
+    let(:body_with_non_default_locale) do
+      AlaveteliLocalization.with_locale('rw') do
+        FactoryBot.create(:public_body, url_name: 'non_default')
+      end
+    end
+
+    context 'when using non-default locale' do
+      it 'fallbacks to the default locale attribute' do
+        AlaveteliLocalization.with_locale('rw') do
+          expect(body_with_default_locale.url_name).to eq 'default'
+        end
+      end
+
+      it 'returns non-default locale attribute' do
+        AlaveteliLocalization.with_locale('rw') do
+          expect(body_with_non_default_locale.url_name).to eq 'non_default'
+        end
+      end
+
+    end
+  end
+
 end
```

**Review by gbp — APPROVED** [review]
Might be good to include the specs in https://github.com/mysociety/alaveteli/pull/5893#issuecomment-698306997

File: lib/alaveteli_localization/locale.rb [hunk 2]
```diff
@@ -49,13 +49,18 @@ def hyphenate
...
 
-    # Note that self_and_parents only uses hyphenated locales
     def self_and_parents
       I18n::Locale::Tag::Simple.new(hyphenate).
         self_and_parents.
         map { |tag| self.class.parse(tag) }
     end
 
+    def i18n_fallbacks(default = nil)
+      default = self.class.parse(default.to_s) if default
```
Comment (garethrees) [inline lib/alaveteli_localization/locale.rb:59]: Redundant `#to_s` – will fix on merge.

**Review by gbp — APPROVED** [review]
:shipit:

**Pull request merged**